### PR TITLE
Travis build

### PR DESCRIPTION
**Pull request by oserban**

### CODE REVIEW

**Comment by senakafdo** [comment]
@oserban, you have two broad options regarding the tests.
1. Uncomment all of those test failures and encourage people to postpone writing tests even further.
2. Do not include `npm test` in the Travis CI build until all apps have the their test cases written.

I personally would like to force people to write tests rather than by-passing this by passing all apps that do not have tests. This is obviously a personal choice, but I think it has its benefits. WDYT?

**Comment by oserban** [comment]
I don't think *code shaming* would work well in this case, because we already have broken build that has been ignored for the last few months. I can't push code that will break the build on every pr, because that means that whoever pushes the first charge on the app repo will have to implement all the tests. I would leave it like this and ask app authors to implement the tests when they are ready.

**Comment by oserban** [comment]
I still believe that my proposal is better for the project because it allows us to add tests iteratively, for each app.

**Comment by jamesscottbrown** [comment]
> I still believe that my proposal is better for the project because it allows us to add tests iteratively, for each app.

Yes, it would clearly be unreasonable to require someone to create tests for *all of the apps*, rather than adding tests for one of them (which would be an improvement, but not resolve Issue #48).

> In other words, I'm saying, it is ok to tell the world that the `test cases have not been executed on a build and therefore it is not stable` rather than saying `there were no test cases to run therefore it could be stable`. I think the two statements are distinctively different.

Yes, and I think the second statement is more correct: if something is untested, its status *is* unknown.

I think the correct statement to make is that as there are no automated tests *unintended regressions* may have been introduced.

I would avoid the use of *stable* in this context, as I don't think either commonly-use meaning is really relevant: OVE may be *stable* (or not) in the sense of changing slowly regardless of whether it has tests; and to have any confidence that it is *stable* in the sense of not crashing we would need to run it continuously for several days.

**Comment by senakafdo** [comment]
@jamesscottbrown - I have approved based on @oserban choice.

Unit testing does more than preventing unintended regressions (for instance see these [docs from Microsoft](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices))

And, you cannot simply justify that `there were no test cases to run therefore it could be stable` is proper. That is certainly a personal preference and generally _not_ the norm in professional software development. As quality standard is meant to be met in whole and not in part.

The only exceptions are if you were on a pre-release branch or if it were a separate repository concerning a non-core component.

**Comment by jamesscottbrown** [comment]
tl;dr: we should use a different word

> And, to add to that, `stability` and `development activity` are two broadly different things. Something that is rapidly developed may be stable as a result of it meeting a certain quality standard (confirmed either via manual or automated testing or by user acceptance). However, something that is not actively developed (i.e. very slow changes at the present compared to the past) does not always mean the software is stable (see http://www.engr.sjsu.edu/fayad/SoftwareStability/Andrsnw2.html).

The article that you have linked does not actually define *stability*: the term *stable* **is** in practice used to refer to both speed of development and reliability (see e.g. https://bitdepth.thomasrutter.com/2010/04/02/stable-vs-stable-what-stable-means-in-software/), which I why I suggested using a more specific term that has only the intended meaning.

If you mean "has been tested" or "has passed QA" then say that, rather than "stable". Note that software may have both "passed QA" and be "unstable" (in the sense of frequently crashing in actual use) if something has been overlooked in the QA process.

The definition of stability in ISO 9126 is "The capability of the software product to avoid unexpected effects from modifications of the software": this definitely is improved by having regression tests, but saying that a particular version of the software is stable because it has been tested seems like a misapplication of the definition (which refers to the *capability* to avoid unexpected effects from modifications, not the absence of unexpected effects resulting from a specific modification).
